File: src1/apps1/tic_tac_toe_vol2o0/websocks/gui/message_manager/ver1o0.py
# BOF [OA16o3o0g8o0]

class MessageManager():
    """[OA16o3o0g8o0] 〇×ゲーム v2 メッセージ駆動"""

    def __init__(self):
        self.addMessageListenerAsync = {}

    def addMessageListener(self, name, setMessage):
        self.addMessageListenerAsync[name] = setMessage

    async def execute(self, scope, doc_received):
        """クライアントからサーバーへ送られてきた変数を解析し、
        サーバーからクライアントへ送信するメッセージの作成"""

        # ログインしていなければ AnonymousUser
        # user = scope["user"]
        # print(f"[MessageManager execute] user=[{user}]")

        # メッセージ名 (Client to server)
        messageName = doc_received.get("message_name", None)

        if(messageName in self.addMessageListenerAsync):
            response_json = await self.addMessageListenerAsync[messageName].on_message_received(scope, doc_received)
            return response_json

        raise ValueError(
            f"[MessageManager execute] unknown message name: {messageName}")


# EOF [OA16o3o0g8o0]

File: src1/apps1/tic_tac_toe_vol2o0/static/tic_tac_toe_vol2o0/msg/message_receiver/ver1o0.js
// BOF [OA16o3o0g3o0]

/**
 * メッセージ受信器
 */
class MessageReceiver {
    constructor() {
        this._messageListeners = {};
    }

    addMessageListener(name, setMessage) {
        this._messageListeners[name] = setMessage;
    }

    /**
     * サーバーからクライアントへ送られてきたメッセージをセットする関数を返します
     * @returns 関数
     */
    execute(message) {
        // メッセージ名 (Server to client)
        let message_name = message["message_name"];
        console.log(`[MessageReceiver execute] サーバーからのメッセージを受信しました message_name:${message_name}`);

        if (message_name in this._messageListeners) {
            // 実行
            const execute2 = this._messageListeners[message_name];
            execute2(message);
        } else {
            // Undefined behavior
            console.log(`[MessageReceiver execute] ignored. message_name:[${message_name}]`);
        }
    }
}

// EOF [OA16o3o0g3o0]
